ykbh/src/lib.rs

```diff
@@ -0,0 +1,310 @@
+}
+
+impl Object {
+
+    /// Creates a new Object from a `usize`.
+    fn from_usize(ptr: usize) -> Object {
+        let data = ptr.to_be_bytes();
+        Object {
+            data: Vec::from(data)
+        }
+    }
+
+    /// Creates a new Object from a `u8`.
+    fn from_u8(val: u8) -> Object {
+        Object {
+            data: Vec::from([val])
+        }
+    }
+
+    /// Converts the objects data back to a `usize`.
+    fn to_usize(&self) -> usize {
+        usize::from_be_bytes(self.data.as_slice().try_into().expect("Unexpected data length."))
+    }
+
+    /// Given a pointer and size, creates a new object by copying `size` data from `ptr`.
+    fn from_ptr(ptr: usize, size: usize) -> Object {
+        let mut data = Vec::with_capacity(size);
+        unsafe {
+            std::ptr::copy(ptr as *const u8, data.as_mut_ptr(), size);
+            data.set_len(size);
+        }
+        Object {
+            data
+        }
+    }
+}
+
+pub struct SIRInterpreter {
+    vars: Vec<Option<Object>>,
+}
+
+impl SIRInterpreter {
+    pub fn new(vars: Vec<Option<Object>>) -> Self {
+        SIRInterpreter {
+            vars,
+        }
+    }
+
+    pub fn interpret(&mut self, body: &ykpack::Body) {
+        // Ignore yktrace::trace_debug.
+        if body.flags & ykpack::bodyflags::TRACE_DEBUG != 0 {
+            return;
+        }
+
+        for stmt in body.blocks[0].stmts.iter() {
+            match stmt {
+                Statement::MkRef(dest, src) => self.mkref(dest, src),
+                Statement::DynOffs { .. } => todo!("dynoffs"),
+                Statement::Store(dest, src) => self.store(dest, src),
+                Statement::BinaryOp { .. } => todo!("binop"),
+                Statement::Nop => {},
+                Statement::Unimplemented(_) | Statement::Debug(_) => { todo!("Unimplemented") },
+                Statement::Cast(..) => todo!("cast"),
+                Statement::Call(..) | Statement::StorageDead(_) => unreachable!()
+            }
+        }
+
+        match &body.blocks[0].term {
+            Terminator::Call {
+                operand: _op,
+                args: _args,
+                destination: _dest
+            } => {
+                todo!()
+            }
+            Terminator::Return => {
+            }
+            t => todo!("{}", t)
+        }
+    }
+
+    /// Get the Object referenced by a Local.
+    fn get_var(&self, local: &Local) -> &Option<Object> {
+        &self.vars[local.0 as usize]
+    }
+
+    /// Implements the Store statement.
+    fn store(&mut self, dest: &IPlace, src: &IPlace) {
+        self.store_obj(dest, self.to_obj(src));
+    }
+
+    /// Write an Object to an IPlace. This either creates a new Object if the IPlace is a Val, or
+    /// writes the data to some pointer, if the IPlace is an Indirect.
+    fn store_obj(&mut self, dest: &IPlace, src: Object) {
+        match dest {
+            IPlace::Val { local, off, ty: _ty } => {
+                // Store into a local.
```

Comment:
Interesting. So we don't mutate the existing bytes, but put in new bytes. Is there a reason?

Reply:
This could be a brand new local, in which case there is nothing to mutate. But you are right, that I could add a check here if the local exists, and if so mutate it instead.